src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeRamBuffer.java

```diff
@@ -0,0 +1,172 @@
+
+        long startingBytesUsed = bytesUsed.get();
+        long startingRowIDsBytesUsed = rowIDsBuilder.ramBytesUsed();
+
+        rowIDsBuilder.add(segmentRowId);
+        bytes.append(value);
+
+        if (segmentRowId != lastSegmentRowID)
+        {
+            numRows++;
+            lastSegmentRowID = segmentRowId;
+        }
+
+        numPoints++;
+
+        long rowIDsAllocatedBytes = rowIDsBuilder.ramBytesUsed() - startingRowIDsBytesUsed;
+        long endingBytesAllocated = bytesUsed.addAndGet(rowIDsAllocatedBytes);
+        
+        return endingBytesAllocated - startingBytesUsed;
+    }
+
+    public IntersectingPointValues asPointValues()
+    {
+        ensureOpen();
+        // building packed longs is destructive
+        closed = true;
+        final PackedLongValues rowIDs = rowIDsBuilder.build();
+        return new IntersectingPointValues()
+        {
+            final int[] ords = IntStream.range(0, numPoints).toArray();
+
+            @Override
+            public void getValue(int i, BytesRef bytesRef)
+            {
+                final long offset = (long) packedValue.length * (long) ords[i];
+                bytesRef.length = packedValue.length;
+                bytes.setRawBytesRef(bytesRef, offset);
+            }
+
+            @Override
+            public byte getByteAt(int i, int k)
+            {
+                final long offset = (long) packedValue.length * (long) ords[i] + (long) k;
+
+                return bytes.readByte(offset);
+            }
+
+            @Override
+            public int getDocID(int i)
+            {
+                return Math.toIntExact(rowIDs.get(ords[i]));
+            }
+
+            @Override
+            public void swap(int i, int j)
+            {
+                int tmp = ords[i];
+                ords[i] = ords[j];
+                ords[j] = tmp;
+            }
+
+            @Override
+            public void intersect(IntersectVisitor visitor) throws IOException
+            {
+                final BytesRef scratch = new BytesRef();
+                for (int i = 0; i < numPoints; i++)
+                {
+                    getValue(i, scratch);
+                    assert scratch.length == packedValue.length;
+                    System.arraycopy(scratch.bytes, scratch.offset, packedValue, 0, packedValue.length);
+                    visitor.visit(getDocID(i), packedValue);
+                }
+            }
+
+            @Override
+            public int getBytesPerDimension()
+            {
+                return packedValue.length;
+            }
+
+            @Override
+            public long size()
+            {
+                return numPoints;
+            }
+
+            @Override
+            public int getDocCount()
+            {
+                return numRows;
+            }
+        };
+    }
+
+    private void ensureOpen()
+    {
+        Preconditions.checkState(!closed, "Expected open buffer.");
```

Comment:
super nit: This could probably be an assertion as well, given it would indicate programmer error, but not super important...